midp/text_editor.js

```diff
@@ -9,110 +13,138 @@ var TextEditorProvider = (function() {
+            this.setPosition(this.left, this.top);
+            this.setBackgroundColor(this.backgroundColor);
+            this.setForegroundColor(this.foregroundColor);
         },
 
-        setStyle: function(styleKey, styleValue) {
-            // Set input/textarea elem styles.
-            if (!this.styles) {
-                this.styles = {};
-            }
-
-            this.styles[styleKey] = styleValue;
-            if (this.textEditorElem) {
+        _setStyle: function(styleKey, styleValue) {
+            if (this.visible) {
                 this.textEditorElem.style.setProperty(styleKey, styleValue);
             }
         },
 
-        getStyle:  function(styleKey) {
-            return (this.styles && this.styles[styleKey]) || null;
-        },
-
         focus: function() {
             this.focused = true;
-            this.textEditorElem && this.textEditorElem.focus();
+            return new Promise(function(resolve, reject) {
+                if (currentVisibleEditor !== this ||
+                    document.activeElement === this.textEditorElem) {
+                    resolve()
+                    return;
+                }
+                setTimeout(this.textEditorElem.focus.bind(this.textEditorElem));
+                this.textEditorElem.onfocus = resolve;
+            }.bind(this));
         },
 
         blur: function() {
             this.focused = false;
-            this.textEditorElem && this.textEditorElem.blur();
+            return new Promise(function(resolve, reject) {
+                if (currentVisibleEditor !== this ||
+                    document.activeElement !== this.textEditorElem) {
+                    resolve();
+                    return;
+                }
+                setTimeout(this.textEditorElem.blur.bind(this.textEditorElem));
+                this.textEditorElem.onblur = resolve;
+            }.bind(this));
         },
 
         getVisible: function() {
-            return this.visible || false;
+            return this.visible;
         },
 
         setVisible: function(aVisible) {
+            // Check if we need to show or hide the html editor elements.
+            if ((currentVisibleEditor === this && aVisible) ||
+                (currentVisibleEditor !== this && !aVisible)) {
+                this.visible = aVisible;
+                return;
+            }
+
             this.visible = aVisible;
 
             if (aVisible) {
+                if (currentVisibleEditor) {
+                    currentVisibleEditor.visible = false;
+                }
+                currentVisibleEditor = this;
+            } else {
+                currentVisibleEditor = null;
+            }
+
+            if (aVisible) {
+                // Add the "show" class to set opacity to 0.
                 // Sometimes in Java, setVisible() is called after focus(), to make
                 // sure the native input won't lose focus, we change opacity instead
                 // of visibility.
```

Comment:
This code comment is obsolete. We can switch back to using visibility and drop the hacking way of setting opacity and z-index.